mllib/src/main/scala/org/apache/spark/mllib/linalg/Vectors.scala

```diff
@@ -449,6 +461,37 @@ class SparseVector(
   override def toString: String =
     "(%s,%s,%s)".format(size, indices.mkString("[", ",", "]"), values.mkString("[", ",", "]"))
 
+  override def equals(other: Any): Boolean = {
+    other match {
+      case v: SparseVector => {
+        if (this.size != v.size) { return false }
+        val thisValues = this.values
+        val thisIndices = this.indices
+        val thisSize = thisValues.size
+        val otherValues = v.values
+        val otherIndices = v.indices
+        val otherSize = otherValues.size
+
+        var k1 = 0
+        var k2 = 0
+        var allEqual = true
+        while (allEqual) {
+          while (k1 < thisSize && thisValues(k1) == 0) k1 += 1
+          while (k2 < otherSize && otherValues(k2) == 0) k2 += 1
+
+          if (k1 >= thisSize || k2 >= otherSize) {
+            return k1 >= thisSize && k2 >= otherSize // check end alignment
+          }
+          allEqual = thisIndices(k1) == otherIndices(k2) && thisValues(k1) == otherValues(k2)
+          k1 += 1
+          k2 += 1
+        }
+        allEqual
+      }
+      case _ => super.equals(other)
+    }
+  }
+
```

Comment:
You mean a function structured like `sqdist`, not a distance-based function right? yes that sounds good to me.